### PR TITLE
value:  rename rev to rot and add binary version of flip

### PR DESCRIPTION
**Pull request by tcolgate**
Implement vertical rotation over the left most dimension.

### CODE REVIEW

**Comment by tcolgate** [comment]
Not too sure on the function name. Also, the APL rot/vrot support picking the dimension to rotate over, I could adapt the vrotate function to support that if preferred. Actually support it would require updates to the syntax to support dimension selection on binary ops (which I don't think is supported at the moment)

**Comment by robpike** [comment]
The error messages say rot not vrot.

I would like to figure out a way to use proper code review tools for pull requests. I find pull requests almost overwhelming difficult to work with otherwise.

**Comment by shurcooL** [comment]
@robpike Have you read through the documentation on pull requests?

https://help.github.com/articles/reviewing-changes-in-pull-requests/

Perhaps there's something there that would be helpful to make them easier to use. You're able to leave inline comments, and draft multiple comments before sending, just like in Gerrit.

**Comment by tcolgate** [comment]
If your preference is for gerrit, but don't want to use googlesource, then
there's a service called GerritHub. I tried it once, without much luck, but
it may be worth looking at.

**Comment by tcolgate** [comment]
@robpike I've updated the error message

File: testdata/binary_matrix.ivy [hunk 1]
```diff
@@ -570,3 +570,29 @@ x[3 2 1]
 
 'abcde' in 3 4 rho 'hello world'
 	0 0 0 1 1
+
```
Comment (robpike) [inline testdata/binary_matrix.ivy]: Please add a vrot of a vector. It should do nothing, but test that.

File: value/binary.go [hunk 2]
```diff
@@ -1172,6 +1172,32 @@ func init() {
...
+					count, ok := countVec[0].(Int)
+					if !ok {
+						Errorf("vrot: count must be small integer")
+					}
+					return v.(Vector).rotate(int(count))
+				},
+				matrixType: func(c Context, u, v Value) Value {
+					countMat := u.(Matrix)
+					if len(countMat.shape) != 1 || len(countMat.data) != 1 {
+						Errorf("rot: count must be small integer")
```
Comment (robpike) [inline value/binary.go]: this message is still rot not vrot.

**Comment by tcolgate** [comment]
@robpike Updates as requested.

**Comment by robpike** [comment]
Aha, I have the right answer about the names. The mistake is in the original, but let's fix it now.

rev and flip in ivy correspond to ⌽ and ⊖ in APL, and both work as unary and binary operators. I believe the right way to solve this problem is to rename (unary) rev to rot, and then your vrot becomes the binary version of flip.

Make sense?

**Comment by tcolgate** [comment]
@ribpike, yes, makes sense now.

I've pushed the renames, but perhaps you'd prefer the rename in a separate commit within this PR?

**Comment by robpike** [comment]
Yes, that's a good idea. Let's do the rename and then add the new binary for flip.

**Comment by tcolgate** [comment]
@robpike Two separate commits, (`git add --patch` is a wonerful thing!).

File: value/binary.go [hunk 3]
```diff
@@ -1172,6 +1172,32 @@ func init() {
 		},
 
 		{
+			name:      "flip",
+			whichType: atLeastVectorType,
+			fn: [numType]binaryFn{
+				vectorType: func(c Context, u, v Value) Value {
+					countVec := u.(Vector)
+					count, ok := countVec[0].(Int)
```
Comment (robpike) [inline value/binary.go]: You need to check that the vector is of length exactly 1.

Comment (tcolgate) [inline value/binary.go]: What would the correct error be? I tried a couple of things on try apl.org. 
`1 2 ⊖ 3 3 ⍴ ⍳ 9` gives a length error

`1 2 3 ⊖ 3 3 ⍴ ⍳ 9` gives
`4 8 3
7 2 6
1 5 9`
I can't pretend to fully understand what is going on there.
For the moment, it seems fair to just error on anything that's not a len(1) vector.

Comment (robpike) [inline value/binary.go]: It makes sense to rotate a matrix by a vector of length equal to the rank of the matrix, but we don't need to do that right now. Let's stick to a scalar, which will arrive either as a length 1 vector or a rank 1 matrix. Either way, data will be of length 1.

The error that's already in the code is fine, but the check is insufficient.

Comment (tcolgate) [inline value/binary.go]: I've pushed an update.

File: testdata/unary_bigint.ivy [hunk 4]
```diff
@@ -71,8 +71,8 @@ rho 10000000000
 ,10000000000
 	10000000000
 
-rev 10000000000
+rot 10000000000
 	10000000000
 
 flip 10000000000
-	10000000000
```
Comment (robpike) [inline testdata/unary_bigint.ivy]: no newline at end of file

File: value/binary.go (hunk 3 above)
Comment (robpike) [inline value/binary.go]: You have to check the length before you index. Unlikely to cause problems but it's possible to have a zero-length vector.

It's possible other places make this mistake, but don't worry about them now. I'll file an issue about it.

**Comment by tcolgate** [comment]
@robpike pushed fix for the length check.

**Comment by robpike** [comment]
Thanks for your patience.

**Pull request merged**